bokehjs/src/coffee/document.coffee

```diff
@@ -650,8 +654,13 @@ export class Document
             throw new Error("Cannot apply patch to #{patched_id} which is not in the document")
           patched_obj = @_all_models[patched_id]
           attr = event_json['attr']
-          value = Document._resolve_refs(event_json['new'], old_references, new_references)
-          patched_obj.setv({ "#{attr}" : value })
+          model_type = event_json['model']['type']
+          if attr == 'data' and model_type == 'ColumnDataSource'
+            [data, shapes] = decode_column_data(event_json['new'])
+            patched_obj.setv({_shapes: shapes, data: data}, {setter_id: setter_id})
```

Comment:
Well, it's not quite a lie, the translation from a NumPy array (which is a `Seq(Any)` to this "little object" that represents typed arrays happens at the very edge of the serialization  boundary (after validation of the NumPy array as a `Seq`). That's basically because the wire-format for typed arrays is something users should never ever see or know about. (tho @philippjfr please correct if I have mis-stated anything) Not sure if there is another way to do or express things.